File: .skills/openclaw-skills/skills/blaspat/feelings-framework/tests/js/test_core.js
/**
 * Feelings Framework — JavaScript Tests
 */

import {
  FeelingsEngine,
  InMemory,
  Feeling,
  FEELING_KEYS,
  defaultCalibration,
} from "../../library/js/feelings/core.js";

// ─── Helpers ────────────────────────────────────────────────────────────────

class DummyMemory extends InMemory {
  /** @type {import('../../library/js/feelings/core.js').FeelingsState|null} */
  _data = null;
  load() { return this._data; }
  save(state) { this._data = state; }
}

// ─── Tests ──────────────────────────────────────────────────────────────────

const tests = [];

function assert(condition, message) {
  if (!condition) throw new Error(`FAIL: ${message}`);
}

function test(name, fn) {
  tests.push({ name, fn });
}

// ─── Test Suites ─────────────────────────────────────────────────────────────

test("initial_state", () => {
  const engine = new FeelingsEngine({ agentId: "test_agent" });
  const state = engine.getState();

  assert(state.agentId === "test_agent", "agent_id matches");
  assert(state.mood === 0.0, "initial mood is 0");
  assert(Object.keys(state.feelings).length === 9, "has 9 feelings");
  assert(Object.values(state.feelings).every(v => v >= 0 && v <= 1), "all feelings in 0..1");
});

test("update_trigger_increases_feeling", () => {
  const engine = new FeelingsEngine({ agentId: "test", memory: new DummyMemory() });
  const state = engine.update("user_praised");

  assert(state.feelings.warmth > 0, "warmth increased after praise");
  assert(state.mood > 0, "mood increased after positive trigger");
});

test("frustration_escalates_with_repetition", () => {
  const engine = new FeelingsEngine({ agentId: "test", memory: new DummyMemory() });

  const s1 = engine.update("request_ignored");
  const s2 = engine.update("request_ignored");
  const s3 = engine.update("request_ignored");

  assert(
    s3.feelings.frustration >= s2.feelings.frustration,
    "frustration escalates with repeated triggers"
  );
});

test("dampening_reduces_feelings_over_time", () => {
  const engine = new FeelingsEngine({ agentId: "test", memory: new DummyMemory() });
  engine.update("user_praised");
  const warmAfterPraise = engine.getState().feelings.warmth;

  // Many neutral updates
  for (let i = 0; i < 15; i++) engine.update("session_started");

  const warmAfterNeutral = engine.getState().feelings.warmth;
  assert(warmAfterNeutral < warmAfterPraise, "warmth dampens with neutral updates");
});

test("calibrate_switch_changes_response", () => {
  const engine = new FeelingsEngine({ agentId: "test", memory: new DummyMemory() });

  // Default agent response
  engine.calibrate("default");
  engine._state.feelings.frustration = 0;
  engine.update("request_ignored");
  const frustrationDefault = engine.getState().feelings.frustration;

  // Calm agent calibration
  const calmCal = defaultCalibration();
  calmCal.triggerDeltas.request_ignored = 0.05;
  engine.setCalibration("calm_agent", calmCal);
  engine.calibrate("calm_agent");
  engine._state.feelings.frustration = 0;
  engine.update("request_ignored");
  const frustrationCalm = engine.getState().feelings.frustration;

  assert(frustrationCalm < frustrationDefault, "calm agent reacts less to same trigger");
});

test("respond_returns_all_modifiers", () => {
  const engine = new FeelingsEngine({ agentId: "test", memory: new DummyMemory() });
  const mods = engine.respond();

  assert(typeof mods.warmth === "number", "has warmth modifier");
  assert(typeof mods.restraint === "number", "has restraint modifier");
  assert(typeof mods.curiosity === "number", "has curiosity modifier");
  assert(typeof mods.guard === "number", "has guard modifier");
  assert(typeof mods.energy === "number", "has energy modifier");
  assert(typeof mods.patience === "number", "has patience modifier");
  assert(typeof mods.reach_out === "number", "has reach_out modifier");
  assert(typeof mods.withdraw === "number", "has withdraw modifier");
  assert(typeof mods.engage_deeper === "number", "has engage_deeper modifier");
  assert(typeof mods.play_it_safe === "number", "has play_it_safe modifier");
  assert(typeof mods.persist === "number", "has persist modifier");
  assert(typeof mods.celebrate === "number", "has celebrate modifier");
  assert(typeof mods._mood === "number", "has _mood context");
  assert(typeof mods._frustration === "number", "has _frustration context");
});

test("register_trigger_runtime", () => {
  const engine = new FeelingsEngine({ agentId: "test", memory: new DummyMemory() });
  engine.registerTrigger("custom_event", Feeling.COOLNESS, 0.5);
  engine.update("custom_event");

  assert(engine.getState().feelings.coolness > 0, "custom trigger applied");
});

test("dampen_all", () => {
  const engine = new FeelingsEngine({ agentId: "test", memory: new DummyMemory() });
  engine.update("user_praised");
  engine.update("surprise_bad");
  const warmBefore = engine.getState().feelings.warmth;
  const anxBefore = engine.getState().feelings.anxiety;

  engine.dampenAll(0.1);

  assert(engine.getState().feelings.warmth < warmBefore, "warmth reduced by dampenAll");
  assert(engine.getState().feelings.anxiety < anxBefore, "anxiety reduced by dampenAll");
});

test("reset_feelings", () => {
  const engine = new FeelingsEngine({ agentId: "test", memory: new DummyMemory() });
  engine.update("user_praised");
  engine.update("surprise_bad");

  const anyPositive = Object.values(engine.getState().feelings).some(v => v > 0);
  assert(anyPositive, "has positive feelings before reset");

  engine.resetFeelings();
  const allZero = Object.values(engine.getState().feelings).every(v => v === 0);
  assert(allZero, "all feelings zero after reset");
});

test("save_and_load_persists_state", () => {
  const mem = new DummyMemory();
  const engine = new FeelingsEngine({ agentId: "test_save", memory: mem });

  engine.update("user_praised");
  engine.update("session_started");
  engine.save();

  const engine2 = new FeelingsEngine({ agentId: "test_save", memory: mem });
  const state2 = engine2.getState();

  assert(state2.feelings.warmth > 0, "warmth persisted across load");
  assert(state2.sessionCount === 2, "session count persisted");
});

test("feeling_keys_is_nine", () => {
  assert(FEELING_KEYS.length === 9, "FEELING_KEYS has 9 entries");
});

test("mood_clamps_to_range", () => {
  const engine = new FeelingsEngine({ agentId: "test", memory: new DummyMemory(), moodMin: -1, moodMax: 1 });
  // Apply many negative triggers
  for (let i = 0; i < 20; i++) engine.update("surprise_bad");
  const mood = engine.getState().mood;
  assert(mood >= -1 && mood <= 1, `mood clamped to range: ${mood}`);
});

test("unknown_trigger_noops", () => {
  const engine = new FeelingsEngine({ agentId: "test", memory: new DummyMemory() });
  const before = engine.getState();
  engine.update("this_trigger_does_not_exist");
  const after = engine.getState();

  assert(
    before.feelings.warmth === after.feelings.warmth &&
    before.mood === after.mood,
    "unknown trigger doesn't change state"
  );
});

// ─── Run ─────────────────────────────────────────────────────────────────────

let passed = 0;
let failed = 0;

for (const { name, fn } of tests) {
  try {
    fn();
    console.log(`  ✓ ${name}`);
    passed++;
  } catch (err) {
    console.error(`  ✗ ${name}`);
    console.error(`    ${err.message}`);
    failed++;
  }
}

console.log(`\n${passed} passed, ${failed} failed`);
if (failed > 0) process.exit(1);

File: .skills/openclaw-skills/skills/blaspat/feelings-framework/tests/python/test_core.py
"""
Feelings Framework — Python Tests
"""

import json
import tempfile
from pathlib import Path

import pytest

from feelings import FeelingsEngine, JsonFileMemory, Feeling, Calibration


class DummyMemory:
    """In-memory memory backend for testing."""

    def __init__(self):
        self._data = None

    def load(self):
        return self._data

    def save(self, state: dict) -> None:
        self._data = state


class TestFeelingsEngine:
    def test_initial_state(self):
        engine = FeelingsEngine(agent_id="test_agent")
        state = engine.get_state()

        assert state["agent_id"] == "test_agent"
        assert state["mood"] == 0.0
        assert len(state["feelings"]) == 9
        assert all(0.0 <= v <= 1.0 for v in state["feelings"].values())

    def test_update_trigger(self):
        engine = FeelingsEngine(agent_id="test", memory=DummyMemory())
        state = engine.update("user_praised")

        assert state["feelings"]["warmth"] > 0.0
        assert state["mood"] > 0.0

    def test_same_trigger_escalates(self):
        engine = FeelingsEngine(agent_id="test", memory=DummyMemory())

        s1 = engine.update("request_ignored")
        s2 = engine.update("request_ignored")
        s3 = engine.update("request_ignored")

        # Third trigger should push frustration higher (escalation)
        assert s3["feelings"]["frustration"] >= s2["feelings"]["frustration"]

    def test_dampening(self):
        engine = FeelingsEngine(agent_id="test", memory=DummyMemory())
        engine.update("user_praised")
        state = engine.get_state()
        initial = state["feelings"]["warmth"]

        # Many neutral updates should slowly reduce warmth
        for _ in range(10):
            engine.update("session_started")
        dampened = engine.get_state()["feelings"]["warmth"]

        assert dampened < initial

    def test_calibrate_switch(self):
        engine = FeelingsEngine(agent_id="test", memory=DummyMemory())
        engine.calibrate("default")
        engine.update("request_ignored")
        frustration_after_default = engine.get_state()["feelings"]["frustration"]

        # Add a calmer calibration
        calm_cal = Calibration.default_calibration()
        calm_cal.trigger_deltas["request_ignored"] = 0.05
        engine.set_calibration("calm_agent", calm_cal)
        engine.calibrate("calm_agent")

        # Reset feelings and re-apply
        engine._state.feelings["frustration"] = 0.0
        engine.update("request_ignored")
        frustration_after_calm = engine.get_state()["feelings"]["frustration"]

        assert frustration_after_calm < frustration_after_default

    def test_respond_returns_modifiers(self):
        engine = FeelingsEngine(agent_id="test", memory=DummyMemory())
        modifiers = engine.respond()

        assert "warmth" in modifiers
        assert "restraint" in modifiers
        assert "curiosity" in modifiers
        assert "guard" in modifiers
        assert "energy" in modifiers
        assert "reach_out" in modifiers
        assert "_mood" in modifiers

    def test_register_trigger_runtime(self):
        engine = FeelingsEngine(agent_id="test", memory=DummyMemory())
        engine.register_trigger("custom_event", Feeling.COOLNESS, delta=0.5)
        engine.update("custom_event")

        assert engine.get_state()["feelings"]["coolness"] > 0.0

    def test_save_load(self):
        with tempfile.TemporaryDirectory() as tmpdir:
            path = Path(tmpdir) / "mood.json"
            engine = FeelingsEngine(agent_id="test_save", memory=JsonFileMemory(path))
            engine.update("user_praised")
            engine.update("session_started")
            engine.save()

            engine2 = FeelingsEngine(agent_id="test_save", memory=JsonFileMemory(path))
            state2 = engine2.get_state()

            assert state2["feelings"]["warmth"] > 0.0
            assert state2["session_count"] == 2

    def test_dampen_all(self):
        engine = FeelingsEngine(agent_id="test", memory=DummyMemory())
        engine.update("user_praised")
        engine.update("surprise_bad")
        warm_before = engine.get_state()["feelings"]["warmth"]
        anx_before = engine.get_state()["feelings"]["anxiety"]

        engine.dampen_all(amount=0.1)

        assert engine.get_state()["feelings"]["warmth"] < warm_before
        assert engine.get_state()["feelings"]["anxiety"] < anx_before

    def test_reset_feelings(self):
        engine = FeelingsEngine(agent_id="test", memory=DummyMemory())
        engine.update("user_praised")
        engine.update("surprise_bad")
        assert any(v > 0 for v in engine.get_state()["feelings"].values())

        engine.reset_feelings()
        assert all(v == 0.0 for v in engine.get_state()["feelings"].values())


class TestFeeling:
    def test_all_returns_nine(self):
        assert len(Feeling.all()) == 9

    def test_feeling_values(self):
        values = {f.value for f in Feeling.all()}
        expected = {
            "warmth", "coolness", "interest", "boredom",
            "loneliness", "security", "anxiety", "satisfaction", "frustration",
        }
        assert values == expected


class TestJsonFileMemory:
    def test_save_and_load(self, tmp_path):
        path = tmp_path / "test_mood.json"
        mem = JsonFileMemory(path)

        state = {"agent_id": "test", "mood": 0.5, "feelings": {"warmth": 0.8}}
        mem.save(state)
        loaded = mem.load()

        assert loaded == state

    def test_load_nonexistent(self, tmp_path):
        path = tmp_path / "does_not_exist.json"
        mem = JsonFileMemory(path)
        assert mem.load() is None


if __name__ == "__main__":
    pytest.main([__file__, "-v"])
